### PR TITLE
Allow plan operations to export JSON

### PR DESCRIPTION
**Pull request by mcasperson**
This PR allows the Terraform plan operations to export the results as JSON and capture the JSON blobs as output variables.

New tests have been added to verify the JSON variable extraction process.

### CODE REVIEW

**Comment by mcasperson** [comment]
[Show Terraform Plan step output as structured JSON instead of a single string variable](https://trello.com/c/feAKRfR1/109-show-terraform-plan-step-output-as-structured-json-instead-of-a-single-string-variable)

File: source/Calamari/Behaviours/PlanBehaviour.cs [hunk 1]
```diff
@@ -49,11 +60,64 @@ protected override Task Execute(RunningDeployment deployment, Dictionary<string,
...
             log.SetOutputVariable(TerraformSpecialVariables.Action.Terraform.PlanOutput, results, deployment.Variables);
+        }
 
-            return this.CompletedTask();
+        public void CaptureJsonOutput(RunningDeployment deployment, string results)
+        {
+            var lines = Regex.Split(results, LineEndingRE);
+            for (var index = 0; index < lines.Length; ++index)
+            {
+                var variableName = $"TerraformPlanLine[{index}].JSON";
```
Comment (mcasperson) [inline source/Calamari/Behaviours/PlanBehaviour.cs:99]: This variable format allows subsequent steps to loop over each JSON blob with Octostache, e.g.:

```
#{each output in Octopus.Action[Plan to apply a Terraform template].Output.TerraformPlanLine}
 Write-Host 'JSON Output line #{output}: #{output.JSON}'
#{/each}
```

File: source/Calamari/Behaviours/PlanBehaviour.cs [hunk 2]
```diff
@@ -24,6 +27,13 @@ class PlanBehaviour : TerraformDeployBehaviour
 
         protected virtual string ExtraParameter => "";
 
+        bool IsUsingPlanJSON(RunningDeployment deployment)
+        {
+            return deployment.Variables.GetFlag(TerraformSpecialVariables.Action.Terraform.PlanJsonOutput);
+        }
+
+        string GetOutputParameter(RunningDeployment deployment) => IsUsingPlanJSON(deployment) ? "--json" : "";
```
Comment (geofflamrock) [inline source/Calamari/Behaviours/PlanBehaviour.cs]: Do older versions of Terraform support `--json` as an option for plan? I thought when discussing this with Sea that he mentioned it wasn't supported in older versions. I downloaded terraform 0.11.15 and it had an error when I ran with `--json` saying it wasn't supported. We may need to make this behaviour version dependent with a warning if the version doesn't support it and its selected.

File: source/Calamari/Behaviours/PlanBehaviour.cs [hunk 3]
```diff
@@ -24,6 +27,13 @@ class PlanBehaviour : TerraformDeployBehaviour
 
         protected virtual string ExtraParameter => "";
 
+        bool IsUsingPlanJSON(RunningDeployment deployment)
+        {
+            return deployment.Variables.GetFlag(TerraformSpecialVariables.Action.Terraform.PlanJsonOutput);
```
Comment (geofflamrock) [inline source/Calamari/Behaviours/PlanBehaviour.cs]: I imagine that this method returns false if the variable doesn't exist?

**Review by geofflamrock — COMMENTED** [review]
Changes look good just not sure on older versions of Terraform and their support for `--json` and whether there is more to do here.

File: source/Calamari/Behaviours/PlanBehaviour.cs (hunk 3 above)
Comment (mcasperson) [reply source/Calamari/Behaviours/PlanBehaviour.cs]: `GetFlag` defaults to false if it isn't found. This is the method signature: `bool GetFlag(string variableName, bool defaultValueIfUnset = false);`

File: source/Calamari/Behaviours/PlanBehaviour.cs (hunk 2 above)
Comment (mcasperson) [reply source/Calamari/Behaviours/PlanBehaviour.cs]: 👍 That's a good point. I'll add a check so this option is not enabled for older versions of Terraform.

**Review by geofflamrock — APPROVED** [review]
Changes look good 👍

**Pull request merged**